manual-provision-payg/testsuite/code_gen.py

```diff
@@ -0,0 +1,320 @@
+    except:
+        # file doens't exist
+        pass
+
+    write_to_file("/usr/local/share/eos-payg/eos-payg.conf",
+                  paygo_enabled_rules)
+
+    generate_codes(keygen_output, read_machine_id())
+
+    # clear_current_code = "sudo gdbus call --system -d com.endlessm.Payg1  -o /com/endlessm/Payg1 -m com.endlessm.Payg1.ClearCode"
+    # execute_command(clear_current_code)
+    bus = dbus.SystemBus()
+    service = bus.get_object('com.endlessm.Payg1', '/com/endlessm/Payg1')
+    clear = service.get_dbus_method("ClearCode", 'com.endlessm.Payg1')
+    clear()
+
+
+def read_machine_id():
+    machine_id = None
+    with open("/etc/machine-id", "r") as f:
+        machine_id = f.read()
+        # print(machine_id) eg '0ce890737896474589e7f3dece38cd7b'
+
+    global device_id
+    # device id currently set as first 8 characters of machine_id string
+    device_id = machine_id[0:8].upper()
+    return device_id
+
+
+def write_to_sqlite(test_code_data):
+    connection_obj = sqlite3.connect(cwd+key_code_file_name + '.db')
+    con = connection_obj.cursor()
+
+    # Create table
+    try:
+        con.execute('''CREATE TABLE keystore
+            (device_id text, test_codes text, key text)''')
+    except Exception as e:
+        # table exists
+        pass
+
+    con.execute("INSERT INTO keystore VALUES ('%s','%s','%s')" % (
+        test_code_data["device_id"], json.dumps(test_code_data["test_codes"]), test_code_data["key"]))
+
+    connection_obj.commit()
+    connection_obj.close()
+
+
+def key_gen(key, device_id):
+    "generates 3 test codes to based on installed key. Period currently set to 1h"
+
+    SIGN_WIDTH_BITS = 13
+    COUNTER_WIDTH_BITS = 8
+
+    counter_list = [250, 249, 248]
+    test_code_list = []
+
+    for count in counter_list:
+
+        key_bytes = bytes(key, "utf-8")
+
+        final_code = None
+
+        digest_maker = None
+        digest_maker = hmac.new(key_bytes, None, hashlib.sha1)
+        digest_maker.update(struct.pack("B", period))
+        digest_maker.update(struct.pack("B", count))
+
+        hmac_data = [0]
+        hmac_data = bytearray(digest_maker.digest())
+        # print(base64.b64encode(hmac_data))
+        sign_mask = (1 << SIGN_WIDTH_BITS) - 1
+        sign_result = ((hmac_data[18]) << 8 | (hmac_data[19])) & sign_mask
+
+        code_value = (period << (COUNTER_WIDTH_BITS + SIGN_WIDTH_BITS)
+                      ) | (count << SIGN_WIDTH_BITS) | sign_result
+
+        if len(str(code_value)) < 8:
+            final_code = "0" + str(code_value)
+
+        else:
+            final_code = str(code_value)
+            # print(code_value)
+
+        test_code_list.append(final_code)
+
+    return {"device_id": device_id, "test_codes": test_code_list, "key": key}
+
+
+def write_to_csv(test_code_data):
+    """write test unlock codes to csv file"""
+    fields_exit = False
+
+    try:
+        if 'device_id' in open(cwd+key_code_file_name+".csv").read():
+            fields_exit = True
```

Comment:
"exist", but can you not determine this by whether the file exists and is non-empty?